pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -288,7 +288,15 @@ private SuccessResponse uploadSegment(FormDataMultiPart multiPart, boolean enabl
           throw new UnsupportedOperationException("Unsupported upload type: " + uploadType);
       }
 
-      String rawTableName = segmentMetadata.getTableName();
+      // Fetch raw table name. Try to derive the table name from the header and then from segment metadata
+      String rawTableName;
+      List<String> tableNameHeader = headers.getRequestHeader(CommonConstants.Controller.TABLE_NAME_HTTP_HEADER);
+      if (tableNameHeader != null) {
```

Comment:
We should not be decoding the metadata if headers are present. Also, introduce a config to reject segments without the headers. Over time, the config will be enabled by default.

Reply:
`We should not be decoding the metadata if headers are present. ` I believe the current behavior is that?

As for config, I was thinking of the following steps for deprecating table name from the segment metadata in the future. 

1. File another PR that deprecates `tableName` from segment metadata -> adding deprecated annotation
2. Keep deprecated fileds for 0.2.0 Apache release
3. Remove tableName from the code when we release 0.3.0 Apache release (when we do that, I was planning to make the code reject if header doesn't have the table name)

do you still think that we need the configuration to reject the segment?

Reply:
ah, sorry I did not consider that we need segment metadata for creating segment zk metadata. Yes, your approach seems fine.